jujugui/static/gui/src/app/components/added-services-list/added-services-list.js

```diff
@@ -6,25 +6,77 @@ const React = require('react');
 const shapeup = require('shapeup');
 
 const AddedServicesListItem = require('./item/item');
+const AddedServicesLabel = require('./label/label');
+
+const addToObj = (obj, key, value) => {
+  if (!obj[key]) {
+    obj[key] = [];
+  }
+  obj[key].push(value);
+};
 
 class AddedServicesList extends React.Component {
+
+  /**
+    Generate a new AddedServicesListItem component and return it.
+    @param {Object} service The service instance to fetch data from.
+    @param {Boolean} lastInList Indicator if it's the last item in a list.
+    @returns {Function} The AddedServicesListItem.
+  */
+  _newListItem(service, lastInList) {
+    return (
+      <AddedServicesListItem
+        // We use the 'name' instead of the 'id' here because when a
+        // ghost service is added it uses the ghost id structure instead
+        // of the final deployed service id structure and we want react
+        // to treat them as the same record instead of re-rendering
+        // when they key changes.
+        changeState={this.props.changeState}
+        hovered={service.get('id') === this.props.hoveredId}
+        key={service.get('name')}
+        lastInList={lastInList}
+        ref={'AddedServicesListItem-' + service.get('id')}
+        service={service}
+        serviceModule={this.props.serviceModule} />);
+  }
+
+  /**
+    Generate the list of the labels and applications.
+    @param {Object} services A YUI ModelList of applications.
+    @returns {Array} A list of applications and label components to render.
+  */
   generateItemList(services) {
-    var items = [];
+    const grouped = {};
+    // Group all of the applications by the bundleURL if one exists.
     services.each(service => {
-      items.push(
-        <AddedServicesListItem
-          // We use the 'name' instead of the 'id' here because when a
-          // ghost service is added it uses the ghost id structure instead
-          // of the final deployed service id structure and we want react
-          // to treat them as the same record instead of re-rendering
-          // when they key changes.
-          changeState={this.props.changeState}
-          hovered={service.get('id') === this.props.hoveredId}
-          key={service.get('name')}
-          ref={'AddedServicesListItem-' + service.get('id')}
-          service={service}
-          serviceModule={this.props.serviceModule} />);
+      const bundleURL = service.get('annotations').bundleURL;
+      if (bundleURL) {
+        addToObj(grouped, bundleURL, service);
+        return;
+      }
+      addToObj(grouped, 'solo', service);
```

Comment:
https://media.giphy.com/media/xTiTno00tQF8TvJxhS/giphy.gif
Anyway, to avoid clashes, maybe call it '++solo++' or just '', or similar. In essence, something that cannot be a valid bundle name. And probably use a const for it?

Reply:
Good call, done